### DISCUSSION TITLE
Jacobians?

### DISCUSSION THREAD

**Discussion by ArvinSKushwaha** [cksac/rai#32] [Q&A] [answered]
Hi! Is there a built in way to do general jacobian computations with `rai`? I assume I could do a loop with JVP/VJP, but that's less preferable to me. Thanks!

**Comment by cksac** [comment]
Currently no, I think we can implement jacfwd and jacrev similar to JAX https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#the-implementation-of-jacfwd-and-jacrev

**Reply by ArvinSKushwaha** [reply]
That would be very cool! I'm hoping to understand more about the intrinsics of this crate more because in the past I wanted to develop something very similar in Rust and right now I'm making very cursed bindings to JAX with PyO3 😆

**Comment by cksac** [comment, marked as answer]
I have implemented the `jacrev` using loop first as there has no vmap exists in rai yet.
```rust
#[test]
fn test_jacrev() {
    let device = Cpu;
    let func = |x: &Tensor| x.sin();
    let jac_fn = jacrev(func);
    let x = &Tensor::from_array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0], [2, 4], device);
    let out = jac_fn(x);
    eval((&x, &out)).unwrap();
    println!("{}", x);
    println!("{}", out);
    println!("{:?}", out.shape());
}
```

**Reply by ArvinSKushwaha** [reply]
Thank you so much, that's very awesome! Great library you have here!

**Reply by cksac** [reply]
wlc. added jacfwd and hessian also. Do you mind share your use case or how you will use rai?

**Reply by ArvinSKushwaha** [reply]
I'm doing work with asynchronous coupled reinforcement learning on Physical systems and I need the expressive type system of Rust, but I also need the machine learning power of JAX, from which I use the functional transforms quite heavily. So I hope that one day soon I can do all that I can with JAX in just Rust.
